Valigator.Core/Data.cs

```diff
@@ -117,5 +119,37 @@ private static IDataContainer<TValue> GetErrorDataContainer(IDataContainer<TValu
 
 		public static implicit operator TValue(Data<TValue> data)
 			=> data.Value;
+
+		public override bool Equals(object obj) 
+			=> obj is Data<TValue> data 
+				&& this == data;//if (obj is Data<TValue> data)//	return this == data;//else//	return false;
+
+		public static bool operator ==(Data<TValue> x, Data<TValue> y) 
+			=> x.DataDescriptor.Equals(y.DataDescriptor)
+				&& (x._value.Equals(y._value) 
+					|| (typeof(TValue).IsArray && StructuralComparisons.StructuralEqualityComparer.Equals(x._value, y._value)));	
+
+		public static bool operator !=(Data<TValue> x, Data<TValue> y)
+			=> !x._value.Equals(y._value);
+
+		public override int GetHashCode()
+		{
+			int hashCode = 943777100;
+			var typeParameter = typeof(TValue);
+			if (typeParameter.IsArray || typeParameter.IsEnum)
+			{
```

Comment:
I haven't dug into it too much, but would using this work? https://docs.microsoft.com/en-us/dotnet/api/system.collections.istructuralequatable.gethashcode?view=net-5.0

Something like:
```C#
(_value as IStructuralEquatable).GetHashCode(EqualityComparer<TValue>.Default)
```

Reply:
Hmm, but `IEnumerable` doesn't implement `IStructuralEquatable`, that might be an issue.

Reply:
And apparently I messed up an assumed you had a generic for the type of the array's elements. That isn't true. I'm not sure how you could get the EqualityComparer off the top of my head.

Reply:
You might be able to do something like this because you don't have the element type as a generic:
`(IEqualityComparer)typeof(EqualityComparer<>).MakeGenericType(elementType).GetProperty(nameof(EqualityComparer<int>.Default)).GetValue(this)`
or the expression tree equivalent so you can cache it.

Reply:
Arguably this isn't the best approach as it does use reflection, I was just hoping to find a way that used as much built in functionality as possible.

Reply:
I'm not against it, I put the change in there.  If this is faster than iterating, I'm happy to do it this way - I'm just guessing that it's not.

Reply:
Yeah, doing a quick running time check might be worth it to see which way is better.

Reply:
Looks like I'm wrong. I did 5 runthroughs for the loop vs Comparer ( 100 equalities & 100 inequalities).  The times for the comparer are stable and low (2-3 millisecond difference between runs).  The loop started fairly low, but had higih variability (1-14 millesecond difference between runs).

I'll leave the IEqualityComparer in there.